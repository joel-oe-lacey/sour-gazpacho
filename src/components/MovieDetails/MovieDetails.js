import React, { Component } from 'react';
import './MovieDetails.scss';
import { connect } from 'react-redux';
import { render } from 'enzyme';
import { fetchData } from '../../utils/fetchCalls'
import { Link } from 'react-router-dom';

class MovieDetails extends Component{
  constructor() {
    super()
    this.state = {
      userRating: 0
    }
  }

  addRating = (rating) => {
    const options = {
        method: "POST",
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({ movie_id: this.props.movie.id, rating: parseInt(rating) }),
    }
    fetchData(`https://rancid-tomatillos.herokuapp.com/api/v1/users/${this.props.userId}/ratings`, options)
      .then(response => response.json())
      .then(rating => this.props.ratings.push(rating.rating))
      .then(data => this.setState({ userRating: data }))
  }

  removeRating = (ratings, movie, userId) => {
    let ratingToRemove = ratings.find(rating => rating.movie_id === movie.id)
    let ratingIndex = ratings.indexOf(ratingToRemove)

    const options = {
        method: "DELETE",
        headers: {
            'Content-Type': 'application/json'
        },
    }
    fetchData(`https://rancid-tomatillos.herokuapp.com/api/v1/users/${parseInt(userId)}/ratings/${ratingToRemove.id}`, options)
      .then(() => ratings.splice(ratingIndex, 1))
      .then(() => this.setState({ userRating: 0 }))
  }

  render() {
    let { movie, userId, ratings } = this.props;
    let month;
    const sectionStyle = {
      backgroundImage: `url(${movie.backdrop_path})`
    }
    const formatDate = movie.release_date.split('-')
    for(var i = 0; i < 13; i++) {
      let months = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December']
      if (formatDate[1] === '10') {
        month = 'October'
      } else if (formatDate[1].includes(0)) {
        if(`0${i}` === formatDate[1]) {
          month = months[i - 1]
        }
      } else {
        if (`${i}` === formatDate[1]) {
          month = months[i - 1]
        }
      }
    }

    const mapRating = ratings.find(rating => rating.movie_id === movie.id)

    return (
      <section className="movie-details" style={ sectionStyle }>
        <h1 className='movie-title-detail'>{movie.title}</h1>
        <div className='movie-detail-card'>
          <img src={movie.poster_path} alt='movie poster' className='movie-poster-detail' />
          <h2 className='detail-title'>{movie.title}</h2>
          <h3 className='detail-rating'>Average Rating: {Math.floor(movie.average_rating)}</h3>
          {mapRating && this.props.userId &&
            <section>
              <h3 className='detail-rating'>{`Your Rating: ${mapRating.rating}`}</h3>
              <button onClick={ (event) => this.removeRating(ratings, movie, userId) } >Delete Rating</button>
            </section>}
          {!mapRating && this.props.userId &&
            <select name="rating" id="rating" onChange={(event) => this.addRating(event.target.value)}>
              <option name="rate" value="">Add Rating</option>
              <option name="rate" value="1">1</option>
              <option name="rate" value="2">2</option>
              <option name="rate" value="3">3</option>
              <option name="rate" value="4">4</option>
              <option name="rate" value="5">5</option>
              <option name="rate" value="6">6</option>
              <option name="rate" value="7">7</option>
              <option name="rate" value="8">8</option>
              <option name="rate" value="9">9</option>
              <option name="rate" value="10">10</option>
            </select>}
          <h3 className='detail-release'>Release Date: {month} {formatDate[2]}, {formatDate[0]}</h3>
          <p className='detail-summary'>Overview: {movie.overview}</p>
        </div>
       <div className='detail-btns'>
        <Link className='movie-btn' to={`/movies/${movie.id-1}`}>See Previous Movie</Link>
        <Link className='movie-btn' to='/'>Go Back to All Movies</Link>
        <Link
         className='movie-btn' to={`/movies/${movie.id+1}`}>See Next Movie</Link>
      </div>
      </section>
    )
  }
}

//make a post on rating
//post returns rating, add that rating into store via dispatch
//does component need to be stateful?




const mapStateToProps = state => ({
  userId: state.user.id,
  ratings: state.ratings
})

export default connect(mapStateToProps)(MovieDetails);
