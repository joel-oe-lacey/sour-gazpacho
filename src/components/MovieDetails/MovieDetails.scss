.movie-detail-card {
  background-color: rgba(255, 255, 255, 0.9);
  border: solid 10px #8E0204;
  margin: 10px;
  // height: 500px;
  width: 70%;
}

.movie-details {
  width: 100%;
  background-size: cover;
  background-position: center;
  padding-top: 100px;
  padding-bottom: 200px;
  overflow-y: scroll;
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  align-items: center;
  color: black;
}

.movie-title-detail {
  color: #FFF;
  font-size: 40px;
  margin-bottom: 50px;
  width: 70%;
  background-color: black;
  background: linear-gradient(90deg, rgba(0,0,0,0) 0%, rgba(0,0,0,1) 50%, rgba(0,0,0,0) 100%);
  height: auto;
  padding: 15px;
}

.movie-poster-detail {
  display: block;
  float: left;
  height: 480px;
  margin-right: 20px;
}

.detail-title {
  font-size: 45px;
  margin: 30px;
}

.detail-rating {
  font-size: 25px;
  color: #8E0204;
}

.detail-release {
  font-size: 25px;
  margin: 20px;
}

.detail-summary {
  font-size: 20px;
}

.detail-btns {
  display: flex;
  flex-direction: row;
  width: 70%;
  text-align: center;
  display: flex;
  justify-content: space-between;
}

.movie-btn {
  width: 30%;
  height: 50px;
  background-color: #FEB40D;
  color: black;
  font-size: 20px;
  border: solid black 0px;
  margin-top: 20px;
  text-decoration: none;
  padding-top: 10px;
}

.hidden {
  visibility: hidden;
}

#mid-btn {
  background-color: #8E0204;
  color: #FFF;
}
