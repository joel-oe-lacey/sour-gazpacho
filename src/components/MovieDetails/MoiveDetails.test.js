import React from 'react';
import { MovieDetails, mapStateToProps } from './MovieDetails.js';
import { shallow } from 'enzyme';

describe('MovieDetails', () => {
let wrapper;
let mockMovie = {
  id: 22,
  title: 'Parasite',
  poster_path: 'https://image.some.jpg',
  backdrop_path: 'https://image.another.jpg',
  release_date: '2019-05-30',
  overview: 'All unemployed, Ki-taek\'s family takes peculiar interest in the wealthy and glamorous Parks for their livelihood until they get entangled in an unexpected incident.',
  average_rating: 5
}
let mockRatings = [
     { id: 826, user_id: 21, movie_id: 23, rating: 8 },
     { id: 827, user_id: 21, movie_id: 22, rating: 5 }
   ]

  beforeEach(() => {
    wrapper = shallow(<MovieDetails
        movie={mockMovie}
        ratings={mockRatings}
        // double check with Eric
      />)
  })

  it('should be an instance of movieDetails', () => {
   expect(wrapper).toMatchSnapshot();
 })

 describe('mapStateToProps', () => {
    it('should return an object with a userId and ratings', () => {
      const mockState = {
        movies: [{id: 1, name: 'Parasite'}, {id: 2, name: 'Sonic'}, {id: 3, name: 'Charlies Angels'}],
        user: { id: 21, name: 'Lucy' },
        ratings: [
             { id: 826, user_id: 21, movie_id: 23, rating: 8 },
             { id: 827, user_id: 21, movie_id: 22, rating: 5 }
           ]
      }
      const expected = {
        userId: 21,
        ratings: mockRatings
      }
      const mappedProps = mapStateToProps( mockState );
      expect(mappedProps).toEqual(expected)
    })
  })
})
