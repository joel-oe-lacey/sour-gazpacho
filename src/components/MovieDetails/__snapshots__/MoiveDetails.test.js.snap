// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails should be an instance of movieDetails 1`] = `
<section
  className="movie-details"
  style={
    Object {
      "backgroundImage": "url(https://image.another.jpg)",
    }
  }
>
  <h1
    className="movie-title-detail"
  >
    Parasite
  </h1>
  <div
    className="movie-detail-card"
  >
    <img
      alt="movie poster"
      className="movie-poster-detail"
      src="https://image.some.jpg"
    />
    <h2
      className="detail-title"
    >
      Parasite
    </h2>
    <h3
      className="detail-rating"
    >
      Average Rating: 
      5
    </h3>
    <h3
      className="detail-release"
    >
      Release Date: 
      May
       
      30
      , 
      2019
    </h3>
    <p
      className="detail-summary"
    >
      Overview: 
      All unemployed, Ki-taek's family takes peculiar interest in the wealthy and glamorous Parks for their livelihood until they get entangled in an unexpected incident.
    </p>
  </div>
  <div
    className="detail-btns"
  >
    <Link
      className="movie-btn"
      to="/movies/21"
    >
      See Previous Movie
    </Link>
    <Link
      className="movie-btn"
      to="/"
    >
      Go Back to All Movies
    </Link>
    <Link
      className="movie-btn"
      to="/movies/23"
    >
      See Next Movie
    </Link>
  </div>
</section>
`;
